### DISCUSSION TITLE
Can't JSON templates use the 'ThreadContext'?

### DISCUSSION THREAD

**Discussion by youngledo** [apache/logging-log4j2#2872] [Q&A]
<img width="365" alt="image" src="https://github.com/user-attachments/assets/be7855b6-2798-4581-ad20-cf448fe58880">

I found that the '%X' in the template is invalid. Is there any other way?

**Comment by ppkarwasz** [comment]
> I found that the '%X' in the template is invalid.

Yes, `%<something>` pattern converters are not supported in the values of templates.

Fortunately for each pattern converter, there is usually a specialized resolver. In this case you can use the [`mdc` resolver](https://logging.staged.apache.org/log4j/2.x/manual/json-template-layout.html#map-resolver-template):

```json
"traceId": {
  "$resolver": "mdc",
  "key": "x-my-TraceID"
}
```

**Note**: In case you don't find an appropriate resolver, you can fall back to:

* using the [`pattern` resolver](https://logging.staged.apache.org/log4j/2.x/manual/json-template-layout.html#event-template-resolver-pattern) with the appropriate pattern converter:
  ```json
  "traceId": {
    "$resolver": "pattern",
    "pattern": "%X{x-my-TraceID}"
  }

* using a [`${...}` property substitution expression](https://logging.staged.apache.org/log4j/2.x/manual/json-template-layout.html#property-substitution-in-template) with the appropriate lookup:
   ```json
   "traceId": "${ctx:x-my-TraceID}"
   ```
